hal/src/rtl872x/sleep_hal.cpp

```diff
@@ -428,21 +457,61 @@ class SleepClass {
             PSRAM_datawr = ((temp << 8) | (temp << 24)) | (temps | (temps << 16));
             PSRAM_CTRL_CA_Gen(psram_ca, (BIT11 | BIT0), PSRAM_LINEAR_TYPE, PSRAM_REG_SPACE, PSRAM_WRITE_TRANSACTION);
             PSRAM_CTRL_DPin_Reg(psram_ca, &PSRAM_datawr, PSRAM_WRITE_TRANSACTION);
+
+            // Disable memory access from CPU
+            // PSRAM_DEV->CSR |= BIT_PSRAM_MEM_IDLE;
+            // while(!(PSRAM_DEV->CSR & BIT_PSRAM_MEM_IDLE));
+
+            // Wait for half-sleep entry just in case
+            DelayUs(200);
         } else {
             temp = HAL_READ32(SYSTEM_CTRL_BASE_LP, REG_AON_LDO_CTRL1);
             temp &= ~(BIT_LDO_PSRAM_EN);
             HAL_WRITE32(SYSTEM_CTRL_BASE_LP, REG_AON_LDO_CTRL1, temp);	
         }
+        __ISB();
+        __DSB();
     }
     
-    __attribute__((section(".ram.sleep"), optimize("O0")))
+    __attribute__((section(".ram.sleep")))
     void resumePsram() {
-        /* dummy read to make psram exit half sleep mode */
-        uint32_t temp = HAL_READ32(PSRAM_BASE, 0);
-        (void)temp;
-        /* wait psram exit half sleep mode */
-        DelayUs(100);
-        DCache_Invalidate((uint32_t)PSRAM_BASE, 32);
+        // Disable memory access from CPU
+        PSRAM_DEV->CSR |= BIT_PSRAM_MEM_IDLE;
+        while(!(PSRAM_DEV->CSR & BIT_PSRAM_MEM_IDLE));
+        uint8_t PSRAM_CA[6];
+        // FIXME: No idea how to align addresses properly here to correctly read things out, reading beginning of PSRAM seems to work though
+        PSRAM_CTRL_CA_Gen(PSRAM_CA, 0, PSRAM_LINEAR_TYPE, PSRAM_MEM_SPACE, PSRAM_READ_TRANSACTION);
+        for (int i = 0; i < 10; i++) {
```

Comment:
We will loop 10x attempting to wake PSRAM? I guess thats the best we can do, we dont really have any other recourse if we cannot access PSRAM at this point.

Reply:
I haven't seen it fail once with this approach honestly. We can potentially fail out and reset/panic.